### ISSUE TITLE
Sik-Ho Tang | Review -- DeepCluster: Deep Clustering for Unsupervised Learning of Visual Features.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#132]
Sik-Ho Tang. [Review — DeepCluster: Deep Clustering for Unsupervised Learning of Visual Features](https://sh-tsang.medium.com/review-deepcluster-deep-clustering-for-unsupervised-learning-of-visual-features-446d0abad74c).

**Comment by NorbertZheng** [comment]
## Overview
DeepCluster, K-Mean Clustering to Generate Pseudo-Labels, a Pretext Task for Self-Supervised Learning.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/33190709-3db6-407a-95b5-0beb1db1a9ed)
Illustration of the Proposed DeepCluster.

In this story, Deep Clustering for Unsupervised Learning of Visual Features, DeepCluster, by Facebook AI Research, is reviewed. In this paper:
 - DeepCluster, a clustering method is proposed that jointly learns the parameters of a neural network and the cluster assignments of the resulting features.
 - DeepCluster iteratively
    - groups the features with a standard clustering algorithm, k-means,
    - and uses the subsequent assignments as supervision to update the weights of the network.

This is a paper in 2018 ECCV with over 900 citations.

**Comment by NorbertZheng** [comment]
K-Mean is applied to learned features, thus generating pseudo-label, EM-learning???

**Comment by NorbertZheng** [comment]
## Notations for Supervised Learning
Before talking about DeepCluster, let’s define some notations using supervised learning.

Given a training set $X=\\{x_{1}, x_{2}, ..., x_{N}\\}$ of $N$ images, we want to find a parameter $\theta^{*}$ such that the mapping $f$ produces good general-purpose features.

These parameters are traditionally learned with supervision, i.e. each image $x_{n}$ is associated with a label $y_{n}$ in $\\{0, 1\\}^{k}$. This label represents the image’s membership to one of $k$ possible predefined classes.

A parametrized classifier $g_{W}$ predicts the correct labels on top of the features $f(x_{n})$.

Therefore, the loss function is (Eq. (1)):
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/4be43f84-f838-4ee8-8248-3c088cd07a19)

where $l$ is the multinomial logistic loss.

This cost function is minimized using mini-batch stochastic gradient descent and backpropagation to compute the gradient.

**Comment by NorbertZheng** [comment]
## DeepCluster as Pretext Task in Self-Supervised Learning
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d4638230-9500-495e-886b-b6ab6e67c042)
Top: k-Means Clustering on Vectors Produced by CNN; Bottom: Using the clustering results as psuedo labels for backpropagation.

### DeepCluster Procedures
The idea of this work is to exploit this weak signal to bootstrap the discriminative power of a convnet.

We cluster the output of the convnet and use the subsequent cluster assignments as “pseudo-labels” to optimize Eq. (1). This deep clustering (DeepCluster) approach iteratively learns the features and groups them.

A standard clustering algorithm, $k$-means, is used.

$k$-means takes a set of vectors as input, in our case the features $f(x_{n})$ produced by the convnet, and clusters them into $k$ distinct groups based on a geometric criterion.

More precisely, it jointly learns a $d\times k$ centroid matrix $C$ and the cluster assignments $y_{n}$ of each image $n$ by solving the following problem (Eq. (2)):

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/930a1c2a-033c-4a59-85fa-c018f4f6900a)
Overall, DeepCluster alternates between clustering the features to produce pseudo-labels using Eq. (2) and updating the parameters of the convnet by predicting these pseudo-labels using Eq. (1).

**Comment by NorbertZheng** [comment]
### Avoiding Trivial Solutions
#### Empty Cluster
An optimal decision boundary is to assign all of the inputs to a single cluster. This issue is caused by the absence of mechanisms to prevent from empty clusters.

More precisely, when a cluster becomes empty, a non-empty cluster is randomly selected and its centroid is used with a small random perturbation as the new centroid for the empty cluster. The points are then reassigned belonging to the non-empty cluster to the two resulting clusters.

#### Trivial Parametrization
If the vast majority of images is assigned to a few clusters, the parameters $\theta$ will exclusively discriminate between them.

A strategy to circumvent this issue is to sample images based on a uniform distribution over the classes, or pseudo-labels.

**Comment by NorbertZheng** [comment]
## DeepCluster Analysis
### Normalized Mutual Information (NMI)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/7dde948f-795a-4109-9a56-2a7450955ace)
(a): Evolution of the clustering quality along training epochs; (b): evolution of cluster reassignments at each clustering step; (c): validation mAP classification performance for various choices of k.

Normalized Mutual Information (NMI), is used to measure the performance:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0945644b-5623-49d4-b6bf-d5f48e2d768c)
where $I$ denotes the mutual information and $H$ the entropy.

If the two assignments A and B are independent, the NMI is equal to 0. If one of them is deterministically predictable from the other, the NMI is equal to 1.
 - (a): The dependence between the clusters and the labels increases over time, showing that the learnt features progressively capture information related to object classes.
 - (b): The NMI is increasing, meaning that there are less and less reassignments and the clusters are stabilizing over time.
 - (c): The best performance is obtained with k= 10,000. Given that ImageNet has 1000 classes. **Apparently some amount of over-segmentation is beneficial.**

**Comment by NorbertZheng** [comment]
### Visualizations
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/21547a8c-872d-48bb-9855-eae28ef8c30e)
Filter visualization and top 9 activated images from a subset of 1 million images from YFCC100M.

As expected, deeper layers in the network seem to capture larger textural structures.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/bca38098-36d4-4a3c-bd0f-87a31e6e69bf)
Top 9 activated images from a random subset of 10 millions images from YFCC100M for target filters in **the last convolutional layer**.

The filters on the top row contain information about structures that highly correlate with object classes. The filters on the bottom row seem to trigger on style, like drawings or abstract shapes.

**Comment by NorbertZheng** [comment]
## DeepCluster Performance
### Linear Classification on Activations on ImageNet & Places
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/78cf0fc5-9728-4ce4-b564-f70b0cf98dc3)
Linear classification on ImageNet and Places using activations from the convolutional layers of an AlexNet as features.

#### ImageNet
Model:
 - A linear classifier is trained on top of different frozen convolutional layers.

Results:
 - On ImageNet, DeepCluster outperforms the state of the art from conv2 to conv5 layers by 1−6%. The largest improvement is observed in the conv3 layer.

Finally, the difference of performance between DeepCluster and a supervised [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160) grows significantly on higher layers: at layers conv2-conv3 the difference is only around 4%, but this difference rises to 12.3% at conv5,
 - **marking where the [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160) probably stores most of the class level information.**

If a MLP is trained on the last layer, DeepCluster outperforms the state of the art by 8%.

#### Places
DeepCluster yields conv3-4 features that are comparable to those trained with ImageNet labels.

 - This suggests that when the target task is sufficiently far from the domain covered by ImageNet, labels are less important.

**Comment by NorbertZheng** [comment]
- If the target domain is exactly the same with the source domain, supervise learning is the best method. E.g. the same subject watching different pictures?
 - If the target domain is different from the source domain, there is little gap between supervise learning and un-supervise learning. E.g. different subjects watching the same pictures or different pictures?

**Comment by NorbertZheng** [comment]
### Pascal VOC
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/4e5e0022-b2d8-4cd8-9936-72208860ae0c)
Comparison of the proposed approach to state-of-the-art unsupervised feature learning on classification, detection and segmentation on Pascal VOC.

DeepCluster outperforms previous unsupervised methods, such as [Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff) [13], [Context Encoders](https://sh-tsang.medium.com/review-context-encoders-feature-learning-by-inpainting-bd181e48997) [46], [Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a) [71], [Split-Brain Auto](https://sh-tsang.medium.com/review-split-brain-autoencoders-unsupervised-learning-by-cross-channel-prediction-cebeba644a3c) [72], [Jigsaw Puzzles](https://sh-tsang.medium.com/review-unsupervised-learning-of-visual-representations-by-solving-jigsaw-puzzles-50b14d755004) [42], on all three tasks, in every setting.

 - The improvement with fine-tuning over the state of the art is the largest on semantic segmentation (7.5%).
 - On detection, DeepCluster performs only slightly better than previously published methods. Interestingly, a fine-tuned random network performs comparatively to many unsupervised methods, but performs poorly if only fc6–8 are learned.

**Comment by NorbertZheng** [comment]
### YFCC100M
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/86f043f4-712d-4828-bf8a-f7a5ac670161)
Impact of the training set on the performance of DeepCluster measured on the Pascal VOC transfer tasks.

 - In YFCC100M, object classes are severly unbalanced, **leading to a data distribution less favorable to DeepCluster**.
 - **This experiment validates that DeepCluster is robust to a change of image distribution**, leading to state-of-the-art general-purpose visual features even if this distribution is not favorable to its design.

**Comment by NorbertZheng** [comment]
### [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160) vs [VGGNet](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/9362b377-4773-4692-91ba-ef24b5e7ba89)
Pascal VOC 2007 object detection with [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160) and [VGG](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11)16.

In the previous experiment, [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160) is used. Here a deeper network [VGGNet](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11) is tried.

Training the [VGG](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11)-16 with DeepCluster gives a performance above the state of the art, bringing us to only 1.4% below the supervised topline.

**Comment by NorbertZheng** [comment]
### Image Retrieval
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d0ffb0d5-32e7-4ad4-9d9b-8571d0d32d4c)
mAP on instance-level image retrieval on Oxford and Paris dataset with a [VGG](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11)16.

The above table suggests that image retrieval is a task where the pre-training is essential and studying it as a down-stream task could give further insights about the quality of the features produced by unsupervised approaches.

**Comment by NorbertZheng** [comment]
One of the major issues is that k-mean clustering takes quite plenty of time.

**Comment by NorbertZheng** [comment]
## Reference
 - [2018 ECCV] [DeepCluster] [Deep Clustering for Unsupervised Learning of Visual Features](https://arxiv.org/abs/1807.05520).
